flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/JobMaster.java

```diff
@@ -473,26 +500,50 @@ public CompletableFuture<Acknowledge> cancel(Time timeout) {
     @Override
     public CompletableFuture<Acknowledge> updateTaskExecutionState(
             final TaskExecutionState taskExecutionState) {
-        FlinkException taskExecutionException;
+        checkNotNull(taskExecutionState, "taskExecutionState");
+        // Use the main/caller thread for all updates to make sure they are processed in order.
+        // (MainThreadExecutor i.e., the akka thread pool does not guarantee that)
+        // Only detach for a FAILED state update that is terminal and may perform io heavy labeling.
+        if (ExecutionState.FAILED.equals(taskExecutionState.getExecutionState())) {
+            return labelFailure(taskExecutionState)
+                    .thenApplyAsync(
+                            taskStateWithLabels -> {
+                                try {
+                                    return doUpdateTaskExecutionState(taskStateWithLabels);
+                                } catch (FlinkException e) {
+                                    throw new CompletionException(e);
+                                }
+                            },
+                            getMainThreadExecutor());
+        }
         try {
-            checkNotNull(taskExecutionState, "taskExecutionState");
+            return CompletableFuture.completedFuture(
+                    doUpdateTaskExecutionState(taskExecutionState));
+        } catch (FlinkException e) {
+            return FutureUtils.completedExceptionally(e);
+        }
+    }
 
+    private Acknowledge doUpdateTaskExecutionState(final TaskExecutionState taskExecutionState)
+            throws FlinkException {
+        @Nullable FlinkException taskExecutionException;
+        try {
             if (schedulerNG.updateTaskExecutionState(taskExecutionState)) {
```

Comment:
That's a good catch; the most straightforward way to address this would be to change the listener to go through JobMaster#updateTaskExecutionState instead so we have a shared (semi-asynchronous) code path for all execution state transitions.

`ExecutionDeploymentReconciliationHandler#onMissingDeploymentsOf` seems to have the same problem (and the fix could most likely be along the same lines).

WDYT?

Reply:
Btw, InternalFailuresListener#notifyTaskFailure is already covered as part of FLINK-31891 -- when a TM disconnection happens we need to Release Payload Slot and since the error is not fromSchedulerNg we use the internalTaskFailuresListener: https://github.com/apache/flink/pull/22511/files#diff-d535f910a10f835962b0637e12014068a9727b2152a84223fd9b1bf9c6c074d6R1665

This is not however covering the `onMissingDeploymentsOf` case David mentioned above

Reply:
How about to do failure enrichment in `SchedulerNG#updateTaskExecutionState(TaskExecutionStateTransition)`? 
A common logic can be introduced to be reused by `SchedulerBase` and `AdaptiveScheduler`.

I prefer to not spread the reference of JobMaster to avoid it being coupled with scheduler internals. InternalFailuresListener is also a workaround to support errors which are directly performed on executions, notifying them to schedulers. It should be removed in the future.

Reply:
The intuition behind choosing JobMaster#updateTaskExecutionState over SchedulerNG#updateTaskExecutionState was mixing asynchronous and synchronous code paths. Since JobMaster#updateTaskExecutionState already returns CompletableFuture, it was easier to integrate with and to reason about ordering guarantees.

Reply:
Also, the `JobMaster#updateTaskExecutionState` code path was scheduler agnostic.

Reply:
If the goal is to label all failures, including those happens within JM, given that the labeling can happen asynchronously, I think it's not avoidable to break current synchronous failure handling process into asynchronous ones.
I agree that there is risk to make this kind of change. So maybe we can combine this new asynchronous labeling process with existing asynchronous processes, to avoid introduce unexpected behavior changes.

For example, for DefaultScheduler, we can invoke `FailureEnricherUtils#labelFailure(...)` in `ExecutionFailureHandler#handleFailure(...)` and directly record the result `CompletableFuture<Map<String, String>>` in `FailureHandlingResult`. Later in `DefaultScheduler#restartTasksWithDelay(...)`, we can wait until the future to be completed before invoking `archiveFromFailureHandlingResult(...)`.

Reply:
A new thought just came up to me. Since the failure labeling is not a critical process, it's better that the failure recovery process will not get blocked due to it. So maybe it's better to just record the `CompletableFuture<Map<String, String>>` in `ExceptionHistoryEntry`. When the exception history is requested, the future can be converted into a `Map<String, String>` if it is completed, otherwise it can be converted into an empty map which will be shown as `Label Unknown Yet` on the web UI.

Reply:
> Since the failure labeling is not a critical process

This is something that we've also considered but it is just pushing the problem far down the road. There has been a consensus that people want to use this for implementing custom restart strategies (e.g., you can decide based on the global context of your infrastructure). This has been intentionally pushed out of the Flip, but it will be a natural follow-up that we shouldn't prevent because too many people have asked for it.

Reply:
> There has been a consensus that people want to use this for implementing custom restart strategies 

If we expect failure enricher to be a crucial part, I would doubt whether we should make it to be an asynchronous process. Before the enrichment is done, a failed job cannot recover, which may last for minutes. 

Also considering future custom restart strategy, one open question is it should act in an asynchronous way, or in a synchronous way to simplify implementation? If taking the former option, a simpler implementation("not treating failure labeling a critical process at the moment") may be more friendlier for future development. If taking the latter option, should the failure enrichment also be synchronous?

Therefore I hesitate to complicate the implementation at the moment before the FLIP of custom restart strategy.
The concern was raised by not answered because custom restart strategy was excluded from the FLIP discussion.
https://lists.apache.org/thread/zoyltjb3k4v9zsbczm8yb4zrw659540v

Reply:
Looks like the main decision we have to take here is if the failure-labels going to be used by restart strategies or not -- relying on enrichment for restarts is what makes it crucial.

The existing implementation was [based on the assumption](https://lists.apache.org/thread/tq8yrncg7zqtpc8ddpxrkxfpovs1wkkw) that labels are going to be used by the custom restart strategies in the future. Since we wanted them asynchronous, the less risky way was through existing async calls e.g., `JobMaster#updateTaskExecutionState`, and probably modifying the InternalFailuresListener (rather than changing SchedulerNG update state to async).

Deciding the failure enrichment is crucial enough to be synchronous -- maybe part of `DefaultScheduler#restartTasksWithDelay`--  is also an option. 

However, decoupling failure labels completely from restart strategies sounds like a step back here.

Reply:
Thanks for updating the PR. @pgaref 
I'm actually not against that restart strategies use failure labels. I'm just hope that we can avoid some complication if possible. Totally +1 to the concern about the risk to change some synchronous process into asynchronous ones. But given that it is not avoidable, considering labeling failure from with the JM, I prefer to do it in a common path. Absolutely later we need to carefully review and test it.
Sorry that I did not have time to review it right now. I may take a look tomorrow.